### PR TITLE
Reject ObjectReferences missing key elements.

### PR DESCRIPTION
**Pull request by mattmoor**
Fixes: https://github.com/knative/pkg/issues/150

### CODE REVIEW

**Comment by mattmoor** [comment]
/hold

We should check in OnChanged as well.

**Comment by mattmoor** [comment]
There isn't a particularly natural way to surface failures in `OnChanged`.  There isn't a logger or context to get one from, and it doesn't return an `error`.

File: tracker/enqueue.go [hunk 1]
```diff
@@ -61,6 +62,24 @@ type set map[string]time.Time
...
+	if ref.APIVersion == "" {
+		missingFields = append(missingFields, "APIVersion")
+	}
+	if ref.Kind == "" {
+		missingFields = append(missingFields, "Kind")
+	}
+	if ref.Namespace == "" {
+		missingFields = append(missingFields, "Namespace")
+	}
+	if ref.Name == "" {
```
Comment (n3wscott) [inline tracker/enqueue.go]: I recommend testing like `if errs := validation.IsQualifiedName(ref.Name); len(errs) > 0 {` https://github.com/knative/eventing/blob/56618d44cfa807ba9a6d7ed26faece5c91447e65/pkg/apis/feeds/v1alpha1/feed_validation.go#L45

File: tracker/enqueue.go [hunk 2]
```diff
@@ -61,6 +62,24 @@ type set map[string]time.Time
 
 // Track implements Interface.
 func (i *impl) Track(ref corev1.ObjectReference, obj interface{}) error {
+	var missingFields []string
+	if ref.APIVersion == "" {
```
[the remark this answers is not in the document]
Comment (mattmoor) [reply tracker/enqueue.go]: IIRC `Pod` can just have `v1`

File: tracker/enqueue.go [hunk 3]
```diff
@@ -61,6 +62,24 @@ type set map[string]time.Time
...
+		missingFields = append(missingFields, "APIVersion")
+	}
+	if ref.Kind == "" {
+		missingFields = append(missingFields, "Kind")
+	}
+	if ref.Namespace == "" {
```
[the remark this answers is not in the document]
Comment (mattmoor) [reply tracker/enqueue.go]: We can relax this, if we need to.  I think you folks are most likely to have duck-typed Cluster-scoped resources, so you tell me :)

File: tracker/enqueue.go (hunk 1 above)
Comment (mattmoor) [reply tracker/enqueue.go]: While we place this limitation on Knative resources, this is not a general requirement in K8s, and so we could be asked to track a resource with a name that departs from this requirement.

File: tracker/enqueue.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (n3wscott) [reply tracker/enqueue.go]: we do have a few cluster scoped things...

[the remark this answers is not in the document]
Comment (n3wscott) [reply tracker/enqueue.go]: Also, we tend to use objectref without namespace with the assumption that if it is empty, the namespace is the same as the surrounding object.

[the remark this answers is not in the document]
Comment (mattmoor) [reply tracker/enqueue.go]: i'd expect you to copy it and pass the fully-qualified thing here

File: tracker/enqueue.go (hunk 1 above)
Comment (evankanderson) [reply tracker/enqueue.go]: I think you'd want `IsDNS1123Label` rather than `IsQualifiedName`; the latter accepts strings like `foo.example.com/Bar`

**Pull request merged**